### PR TITLE
OpenAIClient: use LangChain for auto retry

### PR DESCRIPTION
**Pull request by jwayne**
I added a class `OpenAIClient` in `src/client/openai_client.py`. This class has 2 methods:
* `generate(prompt)`: generates a single chat completion via a single call to OpenAI
* `multi_generate(prompts)`: generates multiple chat completions via parallel calls to OpenAI.

All generations via this class have automatic retries built in via LangChain. See [LangChain's `OpenAIChat` class](https://github.com/hwchase17/langchain/blob/9d42621fa4385e519f702b7005d475781033188c/langchain/llms/openai.py#L659) for available params (all of which can be passed directly into `OpenAIChatClient` during initialization.

You can test this new class by running `python client/openai_client.py`.

Finally, I replaced all references to the predecessor `ChatOpenAI`. This affected `1_Results.py` and `reasoner.py`.

### CODE REVIEW

File: src/openai_chat.py
Comment (RomanEngeler1805) [inline src/openai_chat.py:1]: hey josh, any particular reason why you have kept the original script?

File: scripts/pages/1_📚_Results.py [hunk 1]
```diff
@@ -30,17 +27,18 @@
...
 ENV = os.getenv("STREAMLIT_ENV", "local")
 CONTEXT_WINDOW = 1000
 EXCLUDED_PHRASES = ["Keine relevanten Informationen"]
 RATE_LIMIT_GPT4 = 7
 
-reasoner = Reasoner(CONTEXT_WINDOW)
-ranker = Ranker(mode="jaccard", threshold=0.3, stop_words=stop_words_german)
-
+# Need to make this global to avoid an nltk import issue
+# https://stackoverflow.com/a/27437149/1232944
```
Comment (RomanEngeler1805) [inline scripts/pages/1_📚_Results.py:36]: though I still get quite frequently:

  _File "/Users/romanengeler/hackathons/venv-notion/lib/python3.10/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 565, in _run_script
    exec(code, module.__dict__)
  File "/Users/romanengeler/hackathons/LegalResearchAssistant/scripts/pages/1_📚_Results.py", line 18, in <module>
    stop_words_german = stopwords.words("german")
  File "/Users/romanengeler/hackathons/venv-notion/lib/python3.10/site-packages/nltk/corpus/util.py", line 121, in __getattr__
    self.__load()
  File "/Users/romanengeler/hackathons/venv-notion/lib/python3.10/site-packages/nltk/corpus/util.py", line 95, in __load
    args, kwargs = self.__args, self.__kwargs
AttributeError: 'WordListCorpusReader' object has no attribute '_LazyCorpusLoader__args'_

**Review by RomanEngeler1805 — CHANGES_REQUESTED** [review]
Hey Josh, 
nice work on this is the new client for the chat completion endpoint!

- the unittests fail (test_reasoner.py due to the new chat client)
- it would be great to also have a new client for the completion endpoint, as I'm using it to pre-filter the search results. I'll add a notion task board ticket for it

File: src/openai_chat.py
Comment (jwayne) [reply src/openai_chat.py:1]: Accident, I'll remove.

File: src/reasoner.py [hunk 2]
```diff
@@ -147,20 +149,28 @@ def get_answer(
...
             template=self.templates["answer_prompt"],
         )
-        answer_summary = parallelise_api_calls(_api_client, [prompt])
+        client = OpenAIChatClient(
+            model_name="gpt-4",
+            is_testing=self.is_testing,
+        )
+        answer_summary = client.generate(prompt)
         references = [f"{doc['title']}" for k, doc in enumerate(docs.values())]
         return answer_summary[0], concise_answer_docs[0], references
```
[the remark this answers is not in the document]
Comment (jwayne) [reply src/reasoner.py:170]: Good catch. Typically, comments are preferred. The flow we'd use at Opal is:
- engineer submits PR
- reviewer leaves comments
- engineer fixes and merges PR

File: scripts/pages/1_📚_Results.py (hunk 1 above)
Comment (jwayne) [reply scripts/pages/1_📚_Results.py:36]: Ugh, that's a pain. Having `ranker` be initialized in the global context fixed this for me. Not sure why it's still happening for you. I'm not able to replicate it unfortunately..

Any ideas on unusual setup from your side that might be causing this? E.g. you're starting from a fresh env where this isn't already downloaded, you're on a slow connection, etc

Comment (RomanEngeler1805) [reply scripts/pages/1_📚_Results.py:36]: let's leave it for now and observe in the the live (streamlit) env. if it persists there, we can still think about a fix

**Review by RomanEngeler1805 — APPROVED** [review]

**Comment by RomanEngeler1805** [comment]
Hey Josh,
1) I have pushed a new class llm_filter in in src.ranker on Thursday. This class uses the openai api as well. do you want to refactor the class now, or should we close the current PR and then add a new PR for it? (note that this class has unittests as well)
2) the test_reasoner.py unittest fail (python -m unittest test_reasoner.py)

**Pull request merged**